backend/service/topology/querybuilder.go

```diff
@@ -0,0 +1,141 @@
+package topology
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+const (
+	column            = "column."
+	metadata          = "metadata."
+	queryDefaultLimit = 100
+	maxResultLimit    = 1000
+)
+
+func paginatedQueryBuilder(
+	filter *topologyv1.SearchRequest_Filter,
+	sort *topologyv1.SearchRequest_Sort,
+	pageToken string,
+	limit int,
+) (sq.SelectBuilder, error) {
+	queryLimit := queryDefaultLimit
+	if limit > maxResultLimit {
+		return sq.SelectBuilder{}, errors.New("Maximum query limit is 1000")
+	} else if limit > 0 {
+		queryLimit = limit
+	}
+
+	pageNum, err := strconv.Atoi(pageToken)
+	if err != nil {
+		return sq.SelectBuilder{}, err
+	}
+
+	queryOffset := 0
+	if pageNum > 0 {
+		queryOffset = pageNum * limit
+	}
+
+	query := sq.StatementBuilder.PlaceholderFormat(sq.Dollar).
+		Select("id", "data", "metadata").
+		From("topology_cache").
+		Limit(uint64(queryLimit)).
+		Offset(uint64(queryOffset))
+
+	if filter != nil {
+		query, err = filterQueryBuilder(query, filter)
+		if err != nil {
+			return sq.SelectBuilder{}, err
+		}
+	}
+
+	if sort != nil {
+		query = sortQueryBuilder(query, sort)
+	}
+
+	return query, nil
+}
+
+func filterQueryBuilder(query sq.SelectBuilder, f *topologyv1.SearchRequest_Filter) (sq.SelectBuilder, error) {
+	if f == nil {
+		return query, nil
+	}
+
+	if f.Search != nil && len(f.Search.Field) > 0 {
```

Comment:
is this check needed? seems like Search will always have data b/c of the validation rules https://github.com/lyft/clutch/blob/main/api/topology/v1/topology_api.proto#L60. but i could be misreading / this is best-practice

Reply:
Yeah the proto only generates validation for the gateway, but for unit level testing i'll still need to check again. Which is a bit redundant but necessary IMO to validate the function does not panic with malformed / missing data.